rust/kernel/str.rs

```diff
@@ -373,3 +373,29 @@ mod tests {
         assert_eq!(unchecked_str, "🐧");
     }
 }
+
+// Use `usize` to use `saturating_*` functions.
+pub(crate) struct Formatter {
+    pub(crate) buf: usize,
```

Comment:
This would be unsound. The safety comment of copy_nonoverlapping only make sense because the type can be only constructed inside rust_fmt_argument.

Reply:
I actually wrote some tests earlier today and the safety comment of `copy_nonoverlapping` is wrong even when inside `rust_fmt_argument`: if you look at the implementation of `kasprintf`, you'll find that `rust_fmt_argument` can be called with a `NULL` pointer (https://elixir.bootlin.com/linux/v5.16.8/source/lib/kasprintf.c#L22), so it ends up aborting when `debug_assertions` are enabled.

Reply:
Good catch. I didn't expect NULL to be accepted. Regardless, moving it out from rust_fmt_argument would make it more unsound given `Formatter` can be arbitrarily constructed.